PopPUNK/network.py

```diff
@@ -167,83 +194,124 @@ def extractReferences(G, dbOrder, outPrefix, existingRefs = None, threads = 1):
-    if network_update_required:
+        # Each component is independent, so can be multithreaded
+        components = gt.label_components(G)[0].a
+
+        # Turn gt threading off and on again either side of the parallel loop
+        if gt.openmp_enabled():
+            gt.openmp_set_num_threads(1)
+
+        # Cliques are pruned, taking one reference from each, until none remain
+        with Pool(processes=threads) as pool:
+            ref_lists = pool.map(partial(cliquePrune,
+                                            graph=G,
+                                            reference_indices=reference_indices,
+                                            components_list=components),
+                                 set(components))
+        # Returns nested lists, which need to be flattened
+        reference_indices = set([entry for sublist in ref_lists for entry in sublist])
+
+        if gt.openmp_enabled():
+            gt.openmp_set_num_threads(threads)
+
+        # Use a vertex filter to extract the subgraph of refences
+        # as a graphview
+        reference_vertex = G.new_vertex_property('bool')
+        for n, vertex in enumerate(G.vertices()):
+            if n in reference_indices:
+                reference_vertex[vertex] = True
+            else:
+                reference_vertex[vertex] = False
         G_ref = gt.GraphView(G, vfilt = reference_vertex)
         G_ref = gt.Graph(G_ref, prune = True) # https://stackoverflow.com/questions/30839929/graph-tool-graphview-object
 
-    # Order found references as in mash sketch files
-    reference_names = [dbOrder[int(x)] for x in sorted(reference_indices)]
-    refFileName = writeReferences(reference_names, outPrefix)
-    return reference_indices, reference_names, refFileName, G_ref
+        # Find any clusters which are represented by >1 references
+        # This creates a dictionary: cluster_id: set(ref_idx in cluster)
+        clusters_in_full_graph = printClusters(G, dbOrder, printCSV=False)
+        reference_clusters_in_full_graph = defaultdict(set)
+        for reference_index in reference_indices:
+            reference_clusters_in_full_graph[clusters_in_full_graph[dbOrder[reference_index]]].add(reference_index)
+
+        # Calculate the component membership within the reference graph
+        ref_order = [name for idx, name in enumerate(dbOrder) if idx in frozenset(reference_indices)]
+        clusters_in_reference_graph = printClusters(G_ref, ref_order, printCSV=False)
+        # Record the components/clusters the references are in the reference graph
+        # dict: name: ref_cluster
+        reference_clusters_in_reference_graph = {}
+        for reference_name in ref_order:
+            reference_clusters_in_reference_graph[reference_name] = clusters_in_reference_graph[reference_name]
+
+        # Check if multi-reference components have been split as a validation test
+        # First iterate through clusters
```

Comment:
Is this step not needed with the cugraph pruning – are components guaranteed to be connected after the pruning? If not, this needs to be run

Reply:
Yep, something like this is needed, well spotted.